libirods_rule_engine_plugin-audit_amqp.cpp

```diff
@@ -289,86 +322,95 @@ irods::error exec_rule(
         return err;
     }
 
-    json_t* obj = json_object();
-    if( !obj ) {
-        return ERROR(
-                  SYS_MALLOC_ERR,
-                  "json_object() failed");
-    }
+    nlohmann::json json_obj;
+    std::string tmp_msg;
 
-    struct timeval tv;
-    gettimeofday(&tv, NULL);
-    unsigned long time_ms = tv.tv_sec * 1000 + tv.tv_usec / 1000;
-    std::stringstream time_str; time_str << time_ms;
-    json_object_set(
-        obj,
-        "time_stamp",
-        json_string(time_str.str().c_str()));
+    try {
+        struct timeval tv;
+        gettimeofday(&tv, NULL);
+        unsigned long time_ms = tv.tv_sec * 1000 + tv.tv_usec / 1000;
+        json_obj["time_stamp"] = std::to_string(time_ms);
+
+        char host_name[MAX_NAME_LEN];
+        gethostname( host_name, MAX_NAME_LEN );
+        json_obj["hostname"] = host_name;
+
+        pid_t pid = getpid();
+        json_obj["pid"] = std::to_string(pid);
+
+        json_obj["rule_name"] = _rn;
+
+        for( auto itr : _ps ) {
+            // The BytesBuf parameter should not be serialized because this commonly contains
+            // the entirety of the contents of files. These could be very big and cause the
+            // message broker to explode.
+            if (std::type_index(typeid(BytesBuf*)) == std::type_index(itr.type())) {
+                rodsLog(LOG_DEBUG9, "[{}:{}] - skipping serialization of BytesBuf parameter",
+                    __FILE__, __LINE__);
+                continue;
+            }
 
-    char host_name[MAX_NAME_LEN];
-    gethostname( host_name, MAX_NAME_LEN );
-    json_object_set(
-        obj,
-        "hostname",
-        json_string(host_name));
+            // serialize the parameter to a map
+            irods::re_serialization::serialized_parameter_t param;
+            irods::error ret = irods::re_serialization::serialize_parameter(itr, param);
+            if(!ret.ok()) {
+                 rodsLog(
+                     LOG_ERROR,
+                     "unsupported argument for calling re rules from the rule language");
+                 continue;
+            }
 
-    pid_t pid = getpid();
-    std::stringstream pid_str; pid_str << pid;
-    json_object_set(
-        obj,
-        "pid",
-        json_string(pid_str.str().c_str()));
-
-    json_object_set(
-        obj,
-        "rule_name",
-        json_string(_rn.c_str()));
-
-    for( auto itr : _ps ) {
-        // The BytesBuf parameter should not be serialized because this commonly contains
-        // the entirety of the contents of files. These could be very big and cause the
-        // message broker to explode.
-        if (std::type_index(typeid(BytesBuf*)) == std::type_index(itr.type())) {
-            rodsLog(LOG_DEBUG9, "[{}:{}] - skipping serialization of BytesBuf parameter",
-                __FILE__, __LINE__);
-            continue;
-        }
+            for( auto elem : param ) {
```

Comment:
Consider changing this to `for (const auto& elem : param) {` if `elem` is not being modified.